### ISSUE TITLE
Wayland dropping events when blocking in event loop

### ISSUE THREAD

**Issue by chrisduerr** [rust-windowing/winit#1148]
While trying to refactor the glutin update for Alacritty a bit, I've noticed an issue with the way the event loop operates when it is blocked.

On Wayland, when the window is frequently redrawn and the event loop is blocked for short periods of time, it's possible that key release events are dropped, leading to infinite `ReceiveChar` events.

I was able to reproduce this with this minimal examlpe in Sway:

```rust
mod support;

use glutin::event::{Event, WindowEvent};
use glutin::event_loop::{ControlFlow, EventLoop};
use glutin::window::WindowBuilder;
use glutin::ContextBuilder;

fn main() {
    let el = EventLoop::new();
    let wb = WindowBuilder::new().with_title("A fantastic window!");

    let windowed_context =
        ContextBuilder::new().build_windowed(wb, &el).unwrap();

    let windowed_context = unsafe { windowed_context.make_current().unwrap() };

    println!(
        "Pixel format of the window's GL context: {:?}",
        windowed_context.get_pixel_format()
    );

    let gl = support::load(&windowed_context.context());

    el.run(move |event, _, control_flow| {
        println!("{:?}", event);
        *control_flow = ControlFlow::Wait;

        match event {
            Event::LoopDestroyed => return,
            Event::WindowEvent { ref event, .. } => match event {
                WindowEvent::ReceivedCharacter(c) => {
                    std::thread::sleep_ms(10);
                    gl.draw_frame([1.0, 0.5, 0.7, 1.0]);
                    windowed_context.swap_buffers().unwrap();
                }
                WindowEvent::Resized(logical_size) => {
                    let dpi_factor =
                        windowed_context.window().hidpi_factor();
                    windowed_context
                        .resize(logical_size.to_physical(dpi_factor));
                }
                WindowEvent::RedrawRequested => {
                    std::thread::sleep_ms(10);
                    gl.draw_frame([1.0, 0.5, 0.7, 1.0]);
                    windowed_context.swap_buffers().unwrap();
                }
                WindowEvent::CloseRequested => {
                    *control_flow = ControlFlow::Exit
                }
                _ => (),
            },
            _ => (),
        }
    });
}
```

Pinging @vberger since I've already mentioned this on #glutin and this might be relevant to one of the Wayland libraries.

According to @kchibisov, the `sleep_ms(10)` might not be enough depending on the system, he's had to change it to `30` for example. High key repeat rates are also necessary, the `30` ms were with a repeat rate of `60` and my tests were done with the maximum Sway repeat rate (I've just set it to `600`). Even then a reproduction is a bit spurious, though very reliable in my case (I can reproduce in at least 90% of attempts).

**Comment by kchibisov** [comment]
FWIW my output of `WAYLAND_DEBUG=1` with the above diff and `sleep_ms(30)`. The running example was killed manually, since close window from my WM was ignored.

[out.txt](https://github.com/rust-windowing/winit/files/3585952/out.txt)

**Comment by vberger** [comment]
Okay, I've identified the issue.

They way SCTK works currently is that it generates repeat events from a thread until told to stop, which occurs either when the key is released or the keyboard focus is lost.

Now, winit stores these events in an internal buffer in order to dispatch them to the user EventLoop.

This means that while winit is blocking on some heavy task while a key is pressed, this buffer get filled by repeat events from SCTK. And once the blocking is finished, winit starts emptying this buffer. While the buffer is being emptied, it cannot really process anything else. But it reaches the end eventually. If you wait long enough after releasing the repeating key, the stream of `RepeatedCharacted` comes to an end at some point. At least this is what I observe.

So, I'm not clear how would be the preferred way to solve this, but this is what happens. The thing is that having a high repeat rate generates a real lot of events, and if your app is quite slow, they make take some time to be processed.

**Comment by chrisduerr** [comment]
> So, I'm not clear how would be the preferred way to solve this, but this is what happens. The thing is that having a high repeat rate generates a real lot of events, and if your app is quite slow, they make take some time to be processed.

I'd just like to clarify here that I don't think Alacritty is blocking for very long, we're talking under a ms here if what I'm measuring is what's causing this.

> Could you confirm this is the same for you?

I can confirm that the stream stops eventually, though it takes some time. Holding it down for like a second means it keeps going for minutes.

**Comment by vberger** [comment]
The thing is, if SCTK generates repeat events faster than winit processes them, many can accumulate while the key is pressed.

Then, once the key is released, winit needs to process the whole backlog of repeat events before it actually processes the key release event and signals SCTK to stop repeating. But while winit processes the backlog, SCTK keeps enqueuing new repeat events after the they release, leading to this catastrophic increase in the number of repeat events.

I guess running the test in release mode would make the issue far less noticeable, but yeah, if the repeat rate is high and the winit app processes the repeat events more slowly than they are generated, this is bound to happen.

I guess the solution should be that winit considers repeat event "low priority" compared to the others, though I'm not completely sure how we would manage that.

**Comment by kchibisov** [comment]
> I guess running the test in release mode would make the issue far less noticeable, but yeah, if the repeat rate is high and the winit app processes the repeat events more slowly than they are generated, this is bound to happen.

FWIW in release mod things are worse for me. I mean in debug repeat actually stops, but in release it's keep going infinitely.

**Comment by chrisduerr** [comment]
> I guess running the test in release mode would make the issue far less noticeable

Yeah, fyi this was all tested in release mode of Alacritty and winit/glutin for me. It happens quite reliably.

**Comment by vberger** [comment]
Ah well, I was underestimating the slowness of SCTK generating events in debug I guess. That does not change the heart of the issue: winit / Alacritty does not manage to process the repeat events quickly enough them, and it floods the whole event pipeline.

Given the events are generated from an other thread, if their are actually generated *faster* than winit / Alacritty manages to process them, it can indeed block indefinitely (as the queue grows faster than winit empties it)...
